### PR TITLE
OPS-6642: add slack alerts via infra-event-notifier

### PR DESCRIPTION
**Pull request by bmckerry**
This adds slack alerts using the infra-event-notifier package, which then reroutes those messages to eng pipes.
Requires #20 

Unfortunately slack's markdown support is [lacking](https://www.markdownguide.org/tools/slack/), so instead of having a rendered table we just send a raw ASCII table instead:

- New topics:
image

- Updating topics:
image

- error event (actual error text would be where it says `this is an error`):
image

### CODE REVIEW

File: py/parse_and_notify.py [hunk 1]
```diff
@@ -193,9 +272,17 @@ def build(cls, raw_content: Mapping[str, Any]) -> UpdatedTopic:
 
 
 def main():
-    token = os.getenv("DATADOG_API_KEY")
-    assert token is not None, "No Datadog token in DATADOG_API_KEY env var"
-    notifier = Notifier(datadog_api_key=token)
+    dd_token = os.getenv("DATADOG_API_KEY")
+    slack_secret = os.getenv("KAFKA_CONTROL_PLANE_WEBHOOK_SECRET")
```
Comment (fpacifici) [inline py/parse_and_notify.py:276]: Why is one named after the control plane and the other is not ?

File: py/parse_and_notify.py [hunk 2]
```diff
@@ -218,15 +305,18 @@ def main():
             f"{dry_run}Topicctl ran apply on topic {topic_content.name} "
             f"in region {SENTRY_REGION}"
         )
-        text = topic_content.render_table()
-        if len(text) > 3950:
-            text = (
-                "Changes exceed 4000 character limit, "
+        dd_table = topic_content.render_table()
+        slack_table = topic_content.render_slack_msg()
+        if len(dd_table) > 3950 or len(slack_table) > 2950:
```
Comment (fpacifici) [inline py/parse_and_notify.py:310]: What are those conditions for ?

File: py/parse_and_notify.py [hunk 3]
```diff
@@ -218,15 +305,18 @@ def main():
...
+            dd_table = (
+                "Changes exceed character limit, "
                 "check topicctl logs for details on changes"
             )
```
Comment (fpacifici) [inline py/parse_and_notify.py:314]: Please print what you can anyway.

File: py/parse_and_notify.py [hunk 4]
```diff
@@ -54,6 +55,68 @@ def make_row(row: Sequence[str | int | None]) -> str:
...
+
+    # 3 * (len(headers) - 1) because we need to add 3 dashes for each internal
+    # column separator (i.e ' | ')
+    line = ["-" * (sum(max_width) + 3 * (len(headers) - 1))]
+    rows = [make_row(r, max_width) for r in content]
+    table = (
+        f"```{make_row(headers, max_width)}"
+        + f"{make_row(line, max_width)}"
+        + f"{''.join(rows)}```"
+    )
```
Comment (fpacifici) [inline py/parse_and_notify.py:100]: IT seems to me all this code will just work fine in the markdown version. Did you try to generate the markdown table this way as well? You could remove the copy paste.

File: py/parse_and_notify.py [hunk 5]
```diff
@@ -54,6 +55,68 @@ def make_row(row: Sequence[str | int | None]) -> str:
     return f"%%%\n{table}%%%"
 
 
+def make_slack_message(
+    headers: Sequence[str],
+    content: Sequence[Sequence[str | int | None]],
+    error: bool,
+    error_message: str | None,
```
Comment (fpacifici) [inline py/parse_and_notify.py:62]: You do not need to pass the boolean if the error_message is an optional.
```
if error_message is not None:
 ....
```

**Review by fpacifici — CHANGES_REQUESTED** [review]
Please see comments inline

File: py/parse_and_notify.py (hunk 1 above)
Comment (bmckerry) [reply py/parse_and_notify.py:276]: The DD one is an API key, `slack_secret` is the signing secret for the KCP webhook channel in eng-pipes.
I'll add a comment

File: py/parse_and_notify.py (hunk 2 above)
Comment (bmckerry) [reply py/parse_and_notify.py:310]: DD max event length is 4000, slack max message length is 3000. I set them slightly lower here to allow some wiggle room.
I'll add a comment

Comment (fpacifici) [reply py/parse_and_notify.py]: Please make them constant and avoid magic numbers in code

File: py/parse_and_notify.py (hunk 1 above)
Comment (fpacifici) [reply py/parse_and_notify.py]: No, please do not add a comment to explain the naming convention. 
Naming convention are meant so that you do not have to explain them and they serve as a way to explain what you are doing. 

Please leave it as it is.

File: py/parse_and_notify.py [hunk 6]
```diff
@@ -186,16 +233,27 @@ def build(cls, raw_content: Mapping[str, Any]) -> UpdatedTopic:
...
 
 
 def main():
-    token = os.getenv("DATADOG_API_KEY")
-    assert token is not None, "No Datadog token in DATADOG_API_KEY env var"
-    notifier = Notifier(datadog_api_key=token)
+    # API key for datadog
+    dd_token = os.getenv("DATADOG_API_KEY")
+    # Signing secret for KCP eng-pipes webhook
+    slack_secret = os.getenv("KAFKA_CONTROL_PLANE_WEBHOOK_SECRET")
```
Comment (lynnagara) [inline py/parse_and_notify.py:245]: nit: i'm not sure that "kafka control plane" means anything in our codebases, would `TOPICCTL_WEBHOOK_SECRET` be more explicit?

File: py/parse_and_notify.py [hunk 7]
```diff
@@ -186,16 +233,27 @@ def build(cls, raw_content: Mapping[str, Any]) -> UpdatedTopic:
...
         )
 
 
 def main():
-    token = os.getenv("DATADOG_API_KEY")
-    assert token is not None, "No Datadog token in DATADOG_API_KEY env var"
-    notifier = Notifier(datadog_api_key=token)
+    # API key for datadog
+    dd_token = os.getenv("DATADOG_API_KEY")
+    # Signing secret for KCP eng-pipes webhook
```
Comment (fpacifici) [inline py/parse_and_notify.py]: KCP is not an acronym we use.

File: py/parse_and_notify.py [hunk 8]
```diff
@@ -218,15 +276,26 @@ def main():
...
-        if len(text) > 3950:
-            text = (
-                "Changes exceed 4000 character limit, "
-                "check topicctl logs for details on changes"
+        dd_table = topic_content.render_table(Destinations.DATADOG)
+        slack_table = topic_content.render_table(Destinations.SLACK)
+
+        too_long_message = (
+            "Changes exceed character limit, "
+            "check topicctl logs for more details on changes"
```
Comment (fpacifici) [inline py/parse_and_notify.py:281]: Just add `...` 
So you do not waste a lot of characters if you are hitting the max length

File: py/parse_and_notify.py [hunk 9]
```diff
@@ -7,51 +7,101 @@
...
+            "The following changes were still made:"
+            if len(content) > 1
+            else "No changes were made."
+        )
+        if destination == Destinations.DATADOG:
+            error_header = f"# {error_header}\n"
+            error_footer = f"# {error_footer}\n"
+        elif destination == Destinations.SLACK:
+            error_header = f":warning: *{error_header}*\n"
+            error_footer = f":warning: *{error_footer}*\n"
```
Comment (fpacifici) [inline py/parse_and_notify.py:98]: Why do these have to be different ?

File: py/test_parse_and_notify.py [hunk 10]
```diff
@@ -2,65 +2,60 @@
...
 
 @pytest.mark.parametrize("headers, content, expected", TABLE_TESTS)
-def test_markdown_table(
+def test_table(
     headers: Sequence[str],
     content: Sequence[Sequence[str | int | None]],
     expected: str,
 ) -> None:
-    assert make_markdown_table(headers, content, False, None) == expected
+    assert make_table(headers, content, None, Destinations.SLACK) == expected
```
Comment (fpacifici) [inline py/test_parse_and_notify.py:25]: IF you added two different ways this work, you'll have to test both now. This test is meant to test the corner cases. It is not replaced by `test_topicctl` below. That one tests multiple pieces together.

File: py/parse_and_notify.py (hunk 9 above)
Comment (bmckerry) [reply py/parse_and_notify.py:98]: Slack doesn't support markdown h1's, while datadog markdown doesn't seem to support emojis

I feel like the error text needs to stand out, this was the best way I could come up with

**Review by lynnagara — APPROVED** [review]

**Review by fpacifici — APPROVED** [review]
